crates/matrix-sdk-crypto/src/backups/mod.rs

```diff
@@ -156,26 +268,76 @@ impl BackupMachine {
                             "Checking backup auth data for device"
                         );
 
-                        if let Some(device) = device {
-                            if device.verified()
-                                && device.is_signed_by_device(serialized_auth_data.clone()).is_ok()
-                            {
-                                return Ok(true);
-                            }
-                        } else {
-                            trace!(
-                                device_id = device_key_id.device_id().as_str(),
-                                "Device not found, can't check signature"
-                            );
+                        let state =
+                            self.backup_signed_by_device(device_key_id, device, auth_data.clone());
+
+                        result.insert(device_key_id.device_id().to_owned(), state);
+
+                        // Abort the loop if we found a trusted and valid
+                        // signature, unless we should check all of them.
+                        if state.trusted() && !check_all {
+                            break;
                         }
                     }
                 }
             }
+        }
+
+        Ok(result)
+    }
+
+    async fn verify_auth_data_v1(
+        &self,
+        auth_data: MegolmV1AuthData,
+        check_all: bool,
+    ) -> Result<SignatureCheckResult, CryptoStoreError> {
+        trace!(?auth_data, "Verifying backup auth data");
+        let serialized_auth_data = serde_json::to_value(auth_data.clone())?;
+
+        // Check if there's a signature from our own device.
+        let device_signature = self.check_own_device_signature(serialized_auth_data.clone()).await;
+        // Check if there's a signature from our own user identity.
+        let user_identity_signature =
+            self.check_own_identity_signature(serialized_auth_data.clone()).await?;
+
+        // Collect all the other signatures if there isn't already a valid one,
+        // or if we're told to collect all of them anyways.
+        let other_signatures =
+            if !device_signature.trusted() || !user_identity_signature.trusted() || check_all {
```

Comment:
This will calculate `other_signatures` unless *both* `device_signature` and `user_identity_signature` are trusted, but the comment signals that the intent is to do this only if *neither* of them are trusted.

Reply:
Fixed in https://github.com/matrix-org/matrix-rust-sdk/pull/725/commits/53e75ec939f3ac8398a80753f8d547be22cdd8a1.